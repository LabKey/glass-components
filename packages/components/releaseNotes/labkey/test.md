# @labkey/test
Utilities and configurations for running JavaScript tests with LabKey Server.

### version 0.0.1
*Released*: 31 August 2020
* Initial module.
