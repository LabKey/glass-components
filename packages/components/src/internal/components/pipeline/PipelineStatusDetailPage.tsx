import React, { FC } from 'react';

import classNames from 'classnames';

import { Alert, LoadingSpinner, Page, PageHeader, Section } from '../../..';

import { getPipelineStatusDetail } from './actions';
import { PipelineLogEntry, PipelineStatusDetailModel } from './model';

interface JobStatusRowProps {
    label: string;
    value: string;
}

const JobStatusRow: FC<JobStatusRowProps> = ({ label, value }) => (
    <tr className="pipeline-job-status-detail-row">
        <td>{label}:&nbsp;</td>
        <td>{value}</td>
    </tr>
);

interface WithPipeLineStatusModel {
    model: PipelineStatusDetailModel;
}

const JobStatus: FC<WithPipeLineStatusModel> = ({ model }) => (
    <Section title="Status" panelClassName="pipeline-job-status-detail" titleSize="medium">
        <table>
            <tbody>
                <JobStatusRow label="Created" value={model.created} />
                <JobStatusRow label="Status" value={model.status} />
                <JobStatusRow label="Info" value={model.info} />
            </tbody>
        </table>
    </Section>
);

interface LogEntryProps {
    entry: PipelineLogEntry;
}

const LogEntry: FC<LogEntryProps> = ({ entry }) => (
    <tr
        className={classNames('job-status-log-entry', {
            'job-status-log-entry-warn': entry.level.toLowerCase() === 'warn',
            'job-status-log-entry-error':
                entry.level.toLowerCase() === 'error' || entry.level.toLowerCase() === 'danger',
        })}
    >
        <td>{entry.lines}</td>
    </tr>
);

const LogFile: FC<WithPipeLineStatusModel> = ({ model }) => (
    <Section title="Log" panelClassName="pipeline-job-status-log" titleSize="medium">
        <table>
            <tbody>
                {model.logEntries
                    .filter(entry => !!entry.lines)
                    .map((entry, idx) => (
                        <LogEntry key={idx} entry={entry} />
                    ))}
            </tbody>
        </table>
    </Section>
);

interface Props {
    rowId: number;
    interval?: number; // in ms
}

interface State {
    model: PipelineStatusDetailModel;
    error: any;
}

export class PipelineStatusDetailPage extends React.PureComponent<Props, State> {
    private _task: any;

    static defaultProps = {
        interval: 2000,
    };

    constructor(props) {
        super(props);

        this.state = {
            model: new PipelineStatusDetailModel(),
            error: undefined,
        };
    }

    componentDidMount() {
        this.refresh();
    }

    componentWillUnmount() {
        this.stopRefresh();
    }

    refresh = (): void => {
        const model = this.state.model;

        if (!model?.isLoading) {
            this.setState(
                state => ({
                    model: state.model.mutate({ isLoading: true, isLoaded: false }),
                }),
                () => {
                    const offset = this.state.model?.nextOffset ? this.state.model?.nextOffset : 0;
                    const count = this.state.model?.fetchCount ? this.state.model?.fetchCount + 1 : 1;

                    getPipelineStatusDetail(this.props.rowId, offset, count)
                        .then((model: PipelineStatusDetailModel) => {
                            this.setState(() => ({
                                model,
                            }));
                        })
                        .catch(error => {
                            this.setState(() => ({
                                error,
                            }));
                        });
                }
            );
        }

        if (!this.shouldFetchUpdate()) return;

        this._task = setTimeout(this.refresh, this.props.interval);
    };

    shouldFetchUpdate = (): boolean => {
        return !this.state.error && this.state.model?.active !== false;
    };

    stopRefresh = (): void => {
        if (this._task) clearTimeout(this._task);
    };

    render() {
        const { model, error } = this.state;

        if (!model || !model.isLoaded) return <LoadingSpinner />;
        if (error) return <Alert>{error}</Alert>;

        return (
            <Page title="Status Detail" hasHeader>
                <PageHeader title={model.description} />
                <JobStatus model={model} />
                {model.logEntries && <LogFile model={model} />}
            </Page>
        );
    }
}
