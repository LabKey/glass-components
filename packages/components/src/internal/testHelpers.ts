import { ReactElement } from 'react';
import { act } from 'react-dom/test-utils';
import { Map } from 'immutable';
import { mount, MountRendererProps, ReactWrapper } from 'enzyme';
import { LabKey, Query } from '@labkey/api';
import mock, { proxy } from 'xhr-mock';

import {
    initDomainPropertiesMocks,
    initLineageMocks,
    initPipelineStatusDetailsMocks,
    initQueryGridMocks,
    initUserPropsMocks,
} from '../stories/mock';
import { initQueryGridState, QueryInfo, ServerContextProvider } from '..';

import { RowsResponse } from '../public/QueryModel/QueryModelLoader';

import { applyQueryMetadata, handleSelectRowsResponse } from './query/api';
import { bindColumnRenderers } from './renderers';
import { URL_MAPPERS, URLService } from './url/URLResolver';

declare let LABKEY: LabKey;

export function initMockServerContext(context: Partial<LabKey>): void {
    Object.assign(LABKEY, context);
}

/**
 * Initializes the server context and QueryGrid state which is needed in order to run most tests.
 */
export const initUnitTests = (metadata?: Map<string, any>, columnRenderers?: Map<string, any>): void => {
    initMockServerContext({
        container: {
            formats: {
                dateFormat: 'yyyy-MM-dd',
                dateTimeFormat: 'yyyy-MM-dd HH:mm',
                numberFormat: null,
            },
            path: 'testContainer',
            activeModules: ['Core', 'Query'],
        },
        contextPath: 'labkey',
    });
    initQueryGridState(metadata, columnRenderers);
};

/**
 * Use this method in beforeAll() for your jest tests and you'll have full access
 * to all of the same mock API responses we use in storybook.
 */
export function initUnitTestMocks(
    metadata?: Map<string, any>,
    columnRenderers?: Map<string, any>,
    includePipeline?: boolean
): void {
    window['__react-beautiful-dnd-disable-dev-warnings'] = true;
    initUnitTests(metadata, columnRenderers);
    mock.setup();
    initQueryGridMocks();
    initDomainPropertiesMocks();
    initLineageMocks();
    initUserPropsMocks();
    if (includePipeline) {
        initPipelineStatusDetailsMocks();
    }
    mock.use(proxy);
}

export function registerDefaultURLMappers(): void {
    URLService.registerURLMappers(
        ...URL_MAPPERS.ASSAY_MAPPERS,
        ...URL_MAPPERS.DATA_CLASS_MAPPERS,
        ...URL_MAPPERS.SAMPLE_TYPE_MAPPERS,
        ...URL_MAPPERS.LIST_MAPPERS,
        ...URL_MAPPERS.USER_DETAILS_MAPPERS,
        URL_MAPPERS.LOOKUP_MAPPER
    );
}

/**
 * Instantiates a QueryInfo from a captured query details response payload. Cannot be used until you've called
 * initQueryGridState, initUnitTests, or initUnitTestMocks.
 * @param getQueryDetailsResponse: getQueryDetails response object (e.g. imported from
 * test/data/mixtures-getQueryDetails.json)
 */
export const makeQueryInfo = (getQueryDetailsResponse): QueryInfo => {
    const queryInfo = applyQueryMetadata(getQueryDetailsResponse);
    return queryInfo.merge({ columns: bindColumnRenderers(queryInfo.columns) }) as QueryInfo;
};

/**
 * Creates rows and orderedRows objects needed by the QueryModel. Returns a Promise that resolves to an object that
 * looks like: { messages: any, rows: any, orderedRows: string[], rowCount: number }
 * @param getQueryResponse: getQuery Response object (e.g. imported from test/data/mixtures-getQuery.json)
 */
export const makeTestData = (getQueryResponse): RowsResponse => {
    // Hack: need to stringify and parse the query response object because Query.Response modifies the object in place,
    // which causes errors if you try to use the same response object twice.
    const response = new Query.Response(JSON.parse(JSON.stringify(getQueryResponse)));

    const { key, messages, models, orderedModels, rowCount } = handleSelectRowsResponse(response);

    return {
        messages: messages.toJS(),
        orderedRows: orderedModels[key].toArray(),
        rowCount,
        rows: models[key],
    };
};

/**
 * Use this if you're testing a component that requires a wrapping <ServerContextProvider/> to provide context.
 * This test method wraps enzyme's mount() method and provides the wrapping component with "initialContext".
 * With this the returned mounted component will still be the component under test
 * (as opposed to <ServerContextProvider />).
 * @param node The React node to mount
 * @param initialContext The server context to be provided by the wrapping <ServerContextProvider/>
 * @param options Pass through for mount's rendering options
 */
export const mountWithServerContext = (
    node: ReactElement,
    initialContext: any,
    options?: MountRendererProps
): ReactWrapper => {
    return mount(node, {
        wrappingComponent: ServerContextProvider,
        wrappingComponentProps: { initialContext },
        ...options,
    });
};

/**
 * Use this to sleep in the tests. If you make your test methods async you can use "await sleep();" to put your thread
 * to sleep temporarily which will allow async actions in your component to continue.
 * @param ms
 */
export const sleep = (ms = 0): Promise<void> => {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

/**
 * If you're testing a react component that uses hooks to load data use this method to flush promises and transition
 * your component to the next part of the lifecycle. Example:
 *      const wrapper = mount(<MyComponentThatLoadsStuff />);
 *      // Loading state is expected
 *      expect(wrapper.find(LOADING_SPINNER).exists()).toEqual(true);
 *      await waitForLifecycle(wrapper);
 *      // Items have been loaded and rendered
 *      expect(wrapper.find('.item').length).toEqual(4);
 * @param wrapper: enzyme ReactWrapper
 */
export const waitForLifecycle = (wrapper: ReactWrapper): Promise<undefined> => {
    // Wrap in react-dom/utils act so we don't get errors in our test logs
    return act(async () => {
        await sleep();
        wrapper.update();
    });
};
