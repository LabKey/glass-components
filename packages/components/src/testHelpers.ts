import { Map } from 'immutable';
import { LabKey, Query } from '@labkey/api';
import mock, { proxy } from "xhr-mock";
import { initQueryGridState } from './global';
import { initQueryGridMocks, initLineageMocks, initUserPropsMocks } from './stories/mock';
import { QueryInfo } from './components/base/models/QueryInfo';
import { applyQueryMetadata, handle132Response } from './query/api';
import { bindColumnRenderers } from './renderers';
import { RowsResponse } from './QueryModel/QueryModelLoader';

declare let LABKEY: LabKey;

export function initMockServerContext(context: Partial<LabKey>): void {
    Object.assign(LABKEY, context);
}

/**
 * Initializes the server context and QueryGrid state which is needed in order to run most tests.
 */
export const initUnitTests = (metadata?: Map<string, any>, columnRenderers?: Map<string, any>) => {
    initMockServerContext({
        container: {
            formats: {
                dateFormat: 'yyyy-MM-dd',
                dateTimeFormat: 'yyyy-MM-dd HH:mm',
                numberFormat: null,
            },
            path: 'testContainer',
        },
        contextPath: 'labkey',
    });
    initQueryGridState(metadata, columnRenderers);
};

/**
 * Use this method in beforeAll() for your jest tests and you'll have full access
 * to all of the same mock API responses we use in storybook.
 */
export function initUnitTestMocks(metadata?: Map<string, any>, columnRenderers?: Map<string, any>) {
    initUnitTests(metadata, columnRenderers);
    mock.setup();
    initQueryGridMocks();
    initLineageMocks();
    initUserPropsMocks();
    mock.use(proxy);
}

/**
 * Instantiates a QueryInfo from a captured query details response payload. Cannot be used until you've called
 * initQueryGridState, initUnitTests, or initUnitTestMocks.
 * @param getQueryDetailsResponse: getQueryDetails response object (e.g. imported from
 * test/data/mixtures-getQueryDetails.json)
 */
export const makeQueryInfo = (getQueryDetailsResponse): QueryInfo => {
    const queryInfo = applyQueryMetadata(getQueryDetailsResponse);
    return queryInfo.merge({ columns: bindColumnRenderers(queryInfo.columns) }) as QueryInfo;
};

/**
 * Creates rows and orderedRows objects needed by the QueryModel. Returns a Promise that resolves to an object that
 * looks like: { messages: any, rows: any, orderedRows: string[], rowCount: number }
 * @param getQueryResponse: getQuery Response object (e.g. imported from test/data/mixtures-getQuery.json)
 */
export const makeTestData = (getQueryResponse): Promise<RowsResponse> => {
    // Hack: need to stringify and parse the query response object because Query.Response modifies the object in place,
    // which causes errors if you try to use the same response object twice.
    const response = new Query.Response(JSON.parse(JSON.stringify(getQueryResponse)));
    return handle132Response(response).then((resp) => {
        const { messages, models, orderedModels, rowCount } = resp;
        const key = Object.keys(models)[0];
        return {
            messages: messages.toJS(),
            rows: models[key],
            orderedRows: orderedModels[key].toArray(),
            rowCount,
        };
    });
};

/**
 * Use this to sleep in the tests. If you make your test methods async you can use "await sleep();" to put your thread
 * to sleep temporarily which will allow async actions in your component to continue.
 * @param ms
 */
export const sleep = (ms = 0) => {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};
