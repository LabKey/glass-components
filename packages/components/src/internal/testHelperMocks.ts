import { Map } from 'immutable';
import mock, { proxy } from 'xhr-mock';

import {
    initDomainPropertiesMocks,
    initLineageMocks,
    initPipelineStatusDetailsMocks,
    initQueryGridMocks,
    initUserPropsMocks,
} from '../stories/mock';
import { initUnitTests } from "./testHelpers";

/**
 * Use this method in beforeAll() for your jest tests and you'll have full access
 * to all of the same mock API responses we use in storybook.
 */
export function initUnitTestMocks(
    metadata?: Map<string, any>,
    columnRenderers?: Map<string, any>,
    includePipeline?: boolean
): void {
    window['__react-beautiful-dnd-disable-dev-warnings'] = true;
    initUnitTests(metadata, columnRenderers);
    mock.setup();
    initQueryGridMocks();
    initDomainPropertiesMocks();
    initLineageMocks();
    initUserPropsMocks();
    if (includePipeline) {
        initPipelineStatusDetailsMocks();
    }
    mock.use(proxy);
}
