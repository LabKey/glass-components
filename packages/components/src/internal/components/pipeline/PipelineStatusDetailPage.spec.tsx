import React from 'react';

import { mountWithAppServerContext, waitForLifecycle } from '../../testHelpers';
import { initUnitTestMocks } from '../../../test/testHelperMocks';
import { initPipelineStatusDetailsMocks } from '../../../test/mock';

import { PipelineStatusDetailPage } from './PipelineStatusDetailPage';

beforeAll(() => {
    initUnitTestMocks([initPipelineStatusDetailsMocks]);
});

describe('<PipelineStatusDetailPage>', () => {
    test('Completed job, no warn, no error', async () => {
        const wrapper = mountWithAppServerContext(<PipelineStatusDetailPage rowId={1} />, {}, {});
        await waitForLifecycle(wrapper);
        expect(wrapper.debug({ verbose: false })).toMatchSnapshot();
    });

    test('Failed job, with error', async () => {
        const wrapper = mountWithAppServerContext(<PipelineStatusDetailPage rowId={2} />, {}, {});
        await waitForLifecycle(wrapper);
        expect(wrapper.debug({ verbose: false })).toMatchSnapshot();
    });

    test('Running job, with warning', async () => {
        const wrapper = mountWithAppServerContext(<PipelineStatusDetailPage rowId={3} />, {}, {});
        await waitForLifecycle(wrapper);

        expect(wrapper.debug({ verbose: false })).toMatchSnapshot();
    });
});
