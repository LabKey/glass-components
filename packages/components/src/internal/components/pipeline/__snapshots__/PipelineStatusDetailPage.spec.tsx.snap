// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PipelineStatusDetailPage> Completed job, no warn, no error 1`] = `
"<PipelineStatusDetailPage fetchPipelineStatusDetail={[Function: fetch]} rowId={1} interval={2000}>
  <Page title="Status Detail" hasHeader={true} notFound={false}>
    <div className="app-page">
      <PageHeader title="assays - Workflow_Assay01">
        <div className="page-header">
          <h2 className="no-margin-top">
            assays - Workflow_Assay01
          </h2>
          <Notifications>
            <div className="notifications-container" />
          </Notifications>
        </div>
      </PageHeader>
      <JobStatus model={{...}}>
        <Section title="Status" panelClassName="pipeline-job-status-detail" titleSize="medium">
          <div className="g-section">
            <div className="panel panel-content pipeline-job-status-detail">
              <div className="panel-heading panel-content-flex panel-content-medium">
                <div className="panel-content-title-container ">
                  <div className="panel-content-title-medium ">
                    Status
                  </div>
                </div>
              </div>
              <div className="panel-body">
                <table>
                  <tbody>
                    <JobStatusRow label="Created" value="11:16 2021-01-08">
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Created
                          : 
                        </td>
                        <td>
                          11:16 2021-01-08
                        </td>
                      </tr>
                    </JobStatusRow>
                    <JobStatusRow label="Status" value="COMPLETE">
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Status
                          : 
                        </td>
                        <td>
                          COMPLETE
                        </td>
                      </tr>
                    </JobStatusRow>
                    <JobStatusRow label="Info" value={[undefined]}>
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Info
                          : 
                        </td>
                        <td />
                      </tr>
                    </JobStatusRow>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </Section>
      </JobStatus>
      <LogFile model={{...}}>
        <Section title="Log" panelClassName="pipeline-job-status-log" titleSize="medium">
          <div className="g-section">
            <div className="panel panel-content pipeline-job-status-log">
              <div className="panel-heading panel-content-flex panel-content-medium">
                <div className="panel-content-title-container ">
                  <div className="panel-content-title-medium ">
                    Log
                  </div>
                </div>
              </div>
              <div className="panel-body">
                <table>
                  <tbody>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,328 INFO : ----- Start Batch Properties -----
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,330 INFO : ----- End Batch Properties -----
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,331 INFO : ----- Start Run Properties -----
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,331 INFO : 	Uploaded Files:
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,332 INFO : 		* data_2021-01-08_10-50-22-4.xlsx
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,332 INFO : 	Assay ID: [Blank]
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,332 INFO : 	Run Comments: [Blank]
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,333 INFO : 	dateField1: [Blank]
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,333 INFO : ----- End Run Properties -----
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:11,757 INFO : Starting assay upload
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:12,338 DEBUG: No ExperimentDataHandler registered for data file file:///Users/xing/dev/trunk/build/deploy/files/SM_Workflow_UpdateJob_Test/@files/assaydata/data_2021-01-08_10-50-22-4.log, no special loading will be done on this file.
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          08 Jan 2021 11:16:12,727 INFO : Finished assay upload
                        </td>
                      </tr>
                    </LogEntry>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </Section>
      </LogFile>
    </div>
  </Page>
</PipelineStatusDetailPage>"
`;

exports[`<PipelineStatusDetailPage> Failed job, with error 1`] = `
"<PipelineStatusDetailPage fetchPipelineStatusDetail={[Function: fetch]} rowId={2} interval={2000}>
  <Page title="Status Detail" hasHeader={true} notFound={false}>
    <div className="app-page">
      <PageHeader title="SampleType_01 - SampleType_01_00.xlsx">
        <div className="page-header">
          <h2 className="no-margin-top">
            SampleType_01 - SampleType_01_00.xlsx
          </h2>
          <Notifications>
            <div className="notifications-container" />
          </Notifications>
        </div>
      </PageHeader>
      <JobStatus model={{...}}>
        <Section title="Status" panelClassName="pipeline-job-status-detail" titleSize="medium">
          <div className="g-section">
            <div className="panel panel-content pipeline-job-status-detail">
              <div className="panel-heading panel-content-flex panel-content-medium">
                <div className="panel-content-title-container ">
                  <div className="panel-content-title-medium ">
                    Status
                  </div>
                </div>
              </div>
              <div className="panel-body">
                <table>
                  <tbody>
                    <JobStatusRow label="Created" value="13:42 2021-01-13">
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Created
                          : 
                        </td>
                        <td>
                          13:42 2021-01-13
                        </td>
                      </tr>
                    </JobStatusRow>
                    <JobStatusRow label="Status" value="ERROR">
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Status
                          : 
                        </td>
                        <td>
                          ERROR
                        </td>
                      </tr>
                    </JobStatusRow>
                    <JobStatusRow label="Info" value={[undefined]}>
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Info
                          : 
                        </td>
                        <td />
                      </tr>
                    </JobStatusRow>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </Section>
      </JobStatus>
      <LogFile model={{...}}>
        <Section title="Log" panelClassName="pipeline-job-status-log" titleSize="medium">
          <div className="g-section">
            <div className="panel panel-content pipeline-job-status-log">
              <div className="panel-heading panel-content-flex panel-content-medium">
                <div className="panel-content-title-container ">
                  <div className="panel-content-title-medium ">
                    Log
                  </div>
                </div>
              </div>
              <div className="panel-body">
                <table>
                  <tbody>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          14 Jan 2021 09:31:23,636 INFO : Retrying job.
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          14 Jan 2021 09:31:23,637 DEBUG: Database indicates active task ID is null
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          14 Jan 2021 09:31:23,640 DEBUG: Retry details: Old Job ID: 85d9aaa5-3815-1039-9527-87771aa77304
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          14 Jan 2021 09:31:23,652 INFO : Starting importing SampleType_01 - SampleType_01_00.xlsx
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry job-status-log-entry-error">
                        <td>
                          14 Jan 2021 09:31:29,851 ERROR: Import failed: ERROR: duplicate key value violates unique constraint &quot;uq_material_lsid&quot;
                            Detail: Key (lsid)=(urn:lsid:labkey.com:Sample.8.SampleType_01:S1_1) already exists.
                            Where: SQL statement &quot;INSERT INTO exp.material (rowid, lsid, name, cpastype, created, container, createdby, modifiedby, modified, lastindexed, description, objectid)
                          SELECT $1.RowId, $1.lsid, $1.Name, &#39;urn:lsid:labkey.com:SampleSet.Folder-8:SampleType_01&#39;, CURRENT_TIMESTAMP, &#39;6e20520f-8c10-1038-8703-f26082cf2989&#39;, 1005, 1005, CURRENT_TIMESTAMP, $1.LastIndexed, $1.Description, _$objectid$_&quot;
                          PL/pgSQL function temp.fn_df328d95c900e606a17f6d5d8021e973(temp.fn_df328d95c900e606a17f6d5d8021e973type) line 9 at SQL statement
                        </td>
                      </tr>
                    </LogEntry>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </Section>
      </LogFile>
    </div>
  </Page>
</PipelineStatusDetailPage>"
`;

exports[`<PipelineStatusDetailPage> Running job, with warning 1`] = `
"<PipelineStatusDetailPage fetchPipelineStatusDetail={[Function: fetch]} rowId={3} interval={2000}>
  <Page title="Status Detail" hasHeader={true} notFound={false}>
    <div className="app-page">
      <PageHeader title="exp.data - sourceType1 - data_2021-01-10_06-02-08-2.xlsx">
        <div className="page-header">
          <h2 className="no-margin-top">
            exp.data - sourceType1 - data_2021-01-10_06-02-08-2.xlsx
          </h2>
          <Notifications>
            <div className="notifications-container" />
          </Notifications>
        </div>
      </PageHeader>
      <JobStatus model={{...}}>
        <Section title="Status" panelClassName="pipeline-job-status-detail" titleSize="medium">
          <div className="g-section">
            <div className="panel panel-content pipeline-job-status-detail">
              <div className="panel-heading panel-content-flex panel-content-medium">
                <div className="panel-content-title-container ">
                  <div className="panel-content-title-medium ">
                    Status
                  </div>
                </div>
              </div>
              <div className="panel-body">
                <table>
                  <tbody>
                    <JobStatusRow label="Created" value="07:17 2021-01-10">
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Created
                          : 
                        </td>
                        <td>
                          07:17 2021-01-10
                        </td>
                      </tr>
                    </JobStatusRow>
                    <JobStatusRow label="Status" value="COMPLETE">
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Status
                          : 
                        </td>
                        <td>
                          COMPLETE
                        </td>
                      </tr>
                    </JobStatusRow>
                    <JobStatusRow label="Info" value={[undefined]}>
                      <tr className="pipeline-job-status-detail-row">
                        <td>
                          Info
                          : 
                        </td>
                        <td />
                      </tr>
                    </JobStatusRow>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </Section>
      </JobStatus>
      <LogFile model={{...}}>
        <Section title="Log" panelClassName="pipeline-job-status-log" titleSize="medium">
          <div className="g-section">
            <div className="panel panel-content pipeline-job-status-log">
              <div className="panel-heading panel-content-flex panel-content-medium">
                <div className="panel-content-title-container ">
                  <div className="panel-content-title-medium ">
                    Log
                  </div>
                </div>
              </div>
              <div className="panel-body">
                <table>
                  <tbody>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry">
                        <td>
                          10 Jan 2021 07:17:27,112 INFO : Starting importing exp.data - sourceType1 - data_2021-01-10_06-02-08-2.xlsx
                        </td>
                      </tr>
                    </LogEntry>
                    <LogEntry entry={{...}}>
                      <tr className="job-status-log-entry job-status-log-entry-warn">
                        <td>
                          10 Jan 2021 07:17:29,693 WARN : Parent source not found, skipping...
                        </td>
                      </tr>
                    </LogEntry>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </Section>
      </LogFile>
    </div>
  </Page>
</PipelineStatusDetailPage>"
`;
